runtime/src/accounts_db.rs

```diff
@@ -4586,8 +4597,15 @@ impl AccountsDb {
         &self,
         ancestors: &Ancestors,
         pubkey: &Pubkey,
+        ignore_zero_lamports: LoadZeroLamports,
     ) -> Option<(AccountSharedData, Slot)> {
         self.load(ancestors, pubkey, LoadHint::FixedMaxRoot)
+            .filter(|(account, _)| {
+                matches!(
+                    ignore_zero_lamports,
+                    LoadZeroLamports::SomeWithZeroLamportAccount
+                ) || !account.is_zero_lamport()
+            })
```

Comment:
Why filter here and not deeper in call-stack in `AccountsDb::load` or in `AccountsDb::do_load`, or `AccountsDb::do_load_with_populate_read_cache`?

`AccountsDb::load` at least is a public function and this is potentially is adding a backdoor to bypass this filtering.

Also even though this function is returning `None`, seems like they are still wastefully stored in cache.

Reply:
> Also even though this function is returning None, seems like they are still wastefully stored in cache.

If someone wrote a zero lamport account, it is stored in the write cache. Then it is flushed to an append vec.

Later, if someone uses the index to load the account, load goes to the append vec and reads the account data. Whether it is zero lamport or not, this hits the disk. Currently, anything we load is put in the read cache, under the observation that it will likely be read again. Getting the answer that the account DOES exist in accounts db, but is zero lamport could be a useful answer. Possibly getting it more quickly by having the zero lamport account stored in the read cache is also potentially useful for performance.

Whatever the case, it seems to me that choosing to not put zero lamport accounts in the read cache is a tuning/optimization concern. It is likely of minimal impact and is correct either way.

Incidentally, I think the accounts index also stores a bit that indicates that an entry is zero lamport. We could use that on load, but we are not today. I'll make a note of that. Again, likely minimal impact.

Reply:
The main issue here is
> Why filter here and not deeper in call-stack in `AccountsDb::load` or in `AccountsDb::do_load`, or `AccountsDb::do_load_with_populate_read_cache`?
> `AccountsDb::load` at least is a public function and this is potentially is adding a backdoor to bypass this filtering.

But also I believe what is stored in the cache should be consistent with what api returns. That inconsistently can potentially manifest somewhere else as a bug. So it is not a purely performance issue.

Reply:
> this is potentially is adding a backdoor to bypass this filtering.

This parameterization is required until the feature activates, which is a required parameter by the bank. So, I think the backdoor has to exist.

Reply:
The parametrization is not the issue.
The issue are the other api which by pass the filtering.
In particular `AccountsDb::load` which is public.

and I don't see why `AccountsDb::do_load`, or `AccountsDb::do_load_with_populate_read_cache` should not do the filtering. Either of these can be used to bypass the filtering.

Reply:
> But also I believe what is stored in the cache should be consistent with what api returns. That inconsistently can potentially manifest somewhere else as a bug. So it is not a purely performance issue.

Another perspective is what is stored in the cache matches what is stored in the underlying thing being cached. In this case, the cache matches the appendvec it is caching. As a result, I have higher confidence we'll get the right answer, whether the item is in the read cache, the write cache, or the append vec - all contain the zero lamport account today. However, as usual, you are very insightful. I had noted in my personal notes the same question - of why we would read cache zero lamport accounts. It is a grey area with some tradeoffs, but I landed on the side of keeping the cache consistent with what it is caching.

I am attempting to keep this change as localized, well understood, and simple as possible.

Reply:
Please stop focusing on the cache issue.
As I mentioned the issue here is that you are adding the filtering at a wrong place in the call stack.
`do_load_with_populate_read_cache` is invoked from `load_account_into_read_cache`  which is public function.
So that is another possible route which will bypass the filtering.

Reply:
As long as there are possible paths that you may get different results, this is not fixing the consensus issues.

* `AccountsDb::load` is a public function and it is not doing the filtering. So its output can be different in non-deterministic ways.
* `load_account_into_read_cache` is another public interface with the same issue; and there might be more.

So it is _necessary_ to do the filtering deepest in the call-stack to avoid other possible non-deterministic results causing consensus issues.

As it stands the commit is still leaving room for having consensus issues, and I don't think it is the adequate change.